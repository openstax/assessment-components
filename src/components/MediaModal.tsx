import React, { useEffect, useRef } from 'react';
import styled from 'styled-components';

const buttonHeight = 4.2; // rem
const buttonMargin = 0.5; // rem

// tslint:disable-next-line:variable-name
const Overlay = styled.div`
  position: fixed;
  inset: 0;
  background-color: rgba(0, 0, 0, 0.9);
  z-index: 1000;
  display: flex;
`;

// tslint:disable-next-line:variable-name
const ScrollableContent = styled.div`
  background: white;
  max-width: 100vw;
  max-height: calc(100vh - ${(buttonHeight + buttonMargin * 2) * 2}rem);
  overflow: auto;

  > img {
    ${/*
      fix ScrollableContent height issue where it is slightly larger than
      the image and leaves a gap at the bottom */ ''}
    display: block;
  }
`;


// tslint:disable-next-line:variable-name
const FloatingCloseButton = styled.button`
  position: absolute;
  top: -${buttonHeight + buttonMargin}rem;
  right: ${buttonMargin}rem;
  z-index: 10;
  background: none;
  border: none;
  padding: 0;
  cursor: pointer;
  width: ${buttonHeight}rem;
  height: ${buttonHeight}rem;
`;

// tslint:disable-next-line:variable-name
const ContentContainer = styled.div`
  position: relative;
  pointer-events: auto;
`;

// tslint:disable-next-line:variable-name
const ModalWrapper = styled.div`
  position: fixed;
  inset: 0;
  overflow: hidden;
  z-index: 10;
  display: flex;
  justify-content: center;
  align-items: center;
  pointer-events: none;
`;


// tslint:disable-next-line:variable-name
const CloseIcon = () => (
  <svg width='42' height='42' viewBox='0 0 42 42' xmlns='http://www.w3.org/2000/svg'>
    <rect x='1' y='1' width='40' height='40' rx='20' stroke='white' fill='none' />
    <line x1='16' y1='16' x2='26' y2='26' stroke='white' strokeWidth='2' strokeLinecap='round' />
    <line x1='26' y1='16' x2='16' y2='26' stroke='white' strokeWidth='2' strokeLinecap='round' />
  </svg>
);

interface MediaModalProps {
  isOpen: boolean;
  onClose: () => void;
  children: React.ReactNode;
}
// tslint:disable-next-line:variable-name
const MediaModal: React.FC<MediaModalProps> = ({ isOpen, onClose, children }) => {
  const modalRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (!isOpen || !modalRef.current) return;

    const focusableSelectors = [
      'button:not([disabled])',
      '[tabindex]:not([tabindex="-1"])',
    ].join(', ');

    const focusableElements = modalRef.current.querySelectorAll<HTMLElement>(focusableSelectors);
    const firstEl = focusableElements[0];
    const lastEl = focusableElements[focusableElements.length - 1];

    firstEl?.focus();

    const trapFocus = (e: KeyboardEvent) => {
      if (e.key !== 'Tab') return;

      const activeEl = document.activeElement;
      const isShift = e.shiftKey;

      if (isShift && activeEl === firstEl) {
        e.preventDefault();
        lastEl?.focus();
      } else if (!isShift && activeEl === lastEl) {
        e.preventDefault();
        firstEl?.focus();
      }
    };

    document.addEventListener('keydown', trapFocus);
    return () => {
      document.removeEventListener('keydown', trapFocus);
    };
  }, [isOpen]);

  if (!isOpen) return null;

  return (
    <Overlay onClick={onClose}>
      <ModalWrapper aria-modal="true" role="dialog">
        <ContentContainer
          ref={modalRef}
          onClick={(e) => e.stopPropagation()}
          tabIndex={-1}
        >
          <FloatingCloseButton onClick={onClose} aria-label="Close media preview">
            <CloseIcon />
          </FloatingCloseButton>
          <ScrollableContent>{children}</ScrollableContent>
        </ContentContainer>
      </ModalWrapper>
    </Overlay>
  );
};

export default MediaModal;
