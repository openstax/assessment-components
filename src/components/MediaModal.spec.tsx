import renderer, { act } from 'react-test-renderer';
import MediaModal from './MediaModal';

describe('MediaModal', () => {
  const childContent = <div>Test Content</div>;
  const mockClose = jest.fn();

  beforeEach(() => {
    mockClose.mockReset();
  });

  it('does not render when isOpen is false', () => {
    const tree = renderer.create(
      <MediaModal isOpen={false} onClose={mockClose}>
        {childContent}
      </MediaModal>
    ).toJSON();
    expect(tree).toBeNull();
  });

  it('renders correctly when isOpen is true', () => {
    const tree = renderer.create(
      <MediaModal isOpen={true} onClose={mockClose}>
        {childContent}
      </MediaModal>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls onClose when overlay is clicked', () => {
    const component = renderer.create(
      <MediaModal isOpen={true} onClose={mockClose}>
        {childContent}
      </MediaModal>
    );


    // const overlay = component.root.findByProps({ onClick: mockClose });
    const overlay = component.root.findAllByType('div')[0];

    act(() => {
      overlay.props.onClick();
    });

    expect(mockClose).toHaveBeenCalled();
  });

  it('calls onClose when close button is clicked', () => {
    const component = renderer.create(
      <MediaModal isOpen={true} onClose={mockClose}>
        {childContent}
      </MediaModal>
    );

    const closeButton = component.root.findAllByType('button')[0];

    act(() => {
      closeButton.props.onClick();
    });

    expect(mockClose).toHaveBeenCalled();
  });
});
