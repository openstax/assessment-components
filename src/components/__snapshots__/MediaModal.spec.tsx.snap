// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaModal renders correctly when isOpen is true 1`] = `
<div
  className="sc-bczRLJ drFAoC"
  onClick={[MockFunction]}
>
  <div
    aria-modal="true"
    className="sc-eCYdqJ iZUGnZ"
    role="dialog"
  >
    <div
      className="sc-hKMtZM qarJ"
      onClick={[Function]}
      tabIndex={-1}
    >
      <button
        aria-label="Close media preview"
        className="sc-dkzDqf eiZLKs"
        onClick={[MockFunction]}
      >
        <svg
          height="42"
          viewBox="0 0 42 42"
          width="42"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="none"
            height="40"
            rx="20"
            stroke="white"
            width="40"
            x="1"
            y="1"
          />
          <line
            stroke="white"
            strokeLinecap="round"
            strokeWidth="2"
            x1="16"
            x2="26"
            y1="16"
            y2="26"
          />
          <line
            stroke="white"
            strokeLinecap="round"
            strokeWidth="2"
            x1="26"
            x2="16"
            y1="16"
            y2="26"
          />
        </svg>
      </button>
      <div
        className="sc-gsnTZi kRDrws"
      >
        <div>
          Test Content
        </div>
      </div>
    </div>
  </div>
</div>
`;
